### PR TITLE
ScanAsync fix of handling fully complete stream while future is resolved. Refs #25112.

### PR DESCRIPTION
**Pull request by skisel**

### CODE REVIEW

**Comment by skisel** [comment]
Resolves #25112

**Review by ktoso — APPROVED** [review]
LGTM, thanks!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -511,13 +511,10 @@ private[stream] object Collect {
...
 
       override def onUpstreamFinish(): Unit = {
-        if (current == zero) {
-          eventualCurrent.value match {
-            case Some(Success(`zero`)) ⇒
-              // #24036 upstream completed without emitting anything but after zero was emitted downstream
-              completeStage()
-            case _ ⇒ // in all other cases we will get a complete when the future completes
-          }
+        eventualCurrent.value match {
```
Comment (raboof) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: `onRestart`, `current` is reset but `eventualCurrent` is not. Would that lead to problems when combined with this change?

Comment (skisel) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Good point. I was thinking about it, but this specific case was not solved in the previous implementation. I will try to come up with the test to simulate that situation.

Comment (skisel) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: I think I constructed the test case which should not fail. Please review:

```
      val promises = List(Promise[Int](), Promise[Int]())
      promises(0).complete(Success(1))
      promises(1).complete(Failure(TE("bang")))

      val (pub, sub) =
        TestSource.probe[Int]
          .via(Flow[Int].scanAsync(0)((acc, in) ⇒ promises(in).future))
          .withAttributes(ActorAttributes.supervisionStrategy(Supervision.restartingDecider))
          .toMat(TestSink.probe)(Keep.both)
          .run()

      pub.sendNext(0)
      sub.request(10)
      sub.expectNext(0)
      sub.expectNext(1)
      pub.sendNext(1)
      sub.expectError(TE("bang"))
      pub.sendComplete()
      sub.expectComplete()
```

Am I right that it should be running fine?

Comment (skisel) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Please, disregard this test, I've pushed another fix and simplified the code, also covered some more supervision situations with tests.

**Review by johanandren — APPROVED** [review]
LGTM, thanks!

**Pull request merged**